### PR TITLE
add firebase auth

### PR DESCRIPTION
**Pull request by cfredericks**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,4 +15,5 @@
 .cxx
 local.properties
 /gcp/scripts/key.json
-__pycache__/
\ No newline at end of file
+__pycache__/
+google-services.json
```
Comment (cfredericks) [inline .gitignore:19]: you need to go to https://console.firebase.google.com/u/1/project/stoked-depth-428423-j7/overview and click on the android project at the top followed by the gear to go to settings, and then scroll down to download the google-services.json file and save to the app/ directory in this repo for the emulator to work.

link to android project:
image

download json link:
image

File: app/src/main/res/layout/fragment_leaderboard.xml [hunk 2]
```diff
@@ -6,11 +6,25 @@
...
-        android:paddingBottom="4dp" />
+        android:orientation="horizontal">
+        <Spinner
+            android:id="@+id/tournament_spinner"
+            android:layout_width="0dp"
+            android:layout_weight="1"
+            android:layout_gravity=""
+            android:layout_height="wrap_content"
+            android:paddingBottom="4dp" />
+        <ImageButton
```
Comment (cfredericks) [inline app/src/main/res/layout/fragment_leaderboard.xml:20]: add tournament refresh button

File: app/src/main/res/xml/network_security_config.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (cfredericks) [inline app/src/main/res/xml/network_security_config.xml:1]: allow the emulator to communicate with a local app-engine run (i.e. after running `python3 gcp/app-engine/main.py` in the terminal)

File: app/src/main/java/com/golfbetterapp/golfbetter/TokenManager.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import androidx.credentials.CredentialManagerCallback;
+import androidx.credentials.GetCredentialRequest;
+import androidx.credentials.GetCredentialResponse;
+import androidx.credentials.PasswordCredential;
+
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.auth.FirebaseUser;
+
+public class TokenManager {
```
Comment (cfredericks) [inline app/src/main/java/com/golfbetterapp/golfbetter/TokenManager.java:23]: get auth token from firebase

File: app/src/main/java/com/golfbetterapp/golfbetter/Utils.java [hunk 5]
```diff
@@ -9,73 +11,117 @@
...
+  private static <T> void makeApiCall(final URL url, final JsonParser<T> respParser, final Duration timeout, final String idToken, final ApiCallback<T> callback) {
+    try {
+      final HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+      urlConnection.setRequestMethod("GET");
+      urlConnection.setReadTimeout((int) timeout.toMillis());
+      urlConnection.setConnectTimeout((int) timeout.toMillis());
+      //urlConnection.setDoOutput(true); // only for POST
+
+      if (idToken != null) {
+        urlConnection.setRequestProperty("Authorization", "Bearer " + idToken);
```
Comment (cfredericks) [inline app/src/main/java/com/golfbetterapp/golfbetter/Utils.java:117]: attach token to request, if any

File: gcp/app-engine/main.py [hunk 6]
```diff
@@ -74,6 +86,15 @@ def getconn():
 
 @app.route(PUBLIC_API_PREFIX + '/tournaments', methods=['GET'])
 def get_tournaments():
+    decoded_token = None
```
Comment (cfredericks) [inline gcp/app-engine/main.py]: add auth to endpoints -- maybe in another MR we should add this as an decorator or something to automatically apply auth to endpoints

File: app/src/main/java/com/golfbetterapp/golfbetter/LeaderboardFragment.java [hunk 7]
```diff
@@ -45,10 +51,24 @@ public class LeaderboardFragment extends Fragment {
   private List<Tournament> tournaments;
   private TournamentLeaderboard leaderboard;
 
+  private final BroadcastReceiver REFRESH_RECEIVER = new BroadcastReceiver() {
```
Comment (cfredericks) [inline app/src/main/java/com/golfbetterapp/golfbetter/LeaderboardFragment.java:54]: receiver to refresh tournament data when the other activity finishes the login

File: app/src/main/java/com/golfbetterapp/golfbetter/MainActivity.java [hunk 8]
```diff
@@ -193,6 +148,10 @@ public void handleSignIn(final String user) {
     findViewById(R.id.current_user).setVisibility(View.VISIBLE);
     findViewById(R.id.content_main_container).setVisibility(View.VISIBLE);
     ((TextView)findViewById(R.id.current_user)).setText(signedInUser);
+
+    // Signal tournament refresh to LeaderboardFragment
```
Comment (cfredericks) [inline app/src/main/java/com/golfbetterapp/golfbetter/MainActivity.java:152]: trigger tournament load on login success

File: app/src/main/java/com/golfbetterapp/golfbetter/clients/SportsDataIoAppEngineClient.java [hunk 9]
```diff
@@ -1,37 +0,0 @@
-package com.golfbetterapp.golfbetter.clients;
```
Comment (cfredericks) [inline app/src/main/java/com/golfbetterapp/golfbetter/clients/SportsDataIoAppEngineClient.java:1]: as discussed over zoom, delete a bunch of old sportsdataio classes

File: app/src/main/res/drawable/refresh.xml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android" android:height="24dp" android:tint="#000000" android:viewportHeight="24" android:viewportWidth="24" android:width="24dp">
```
Comment (cfredericks) [inline app/src/main/res/drawable/refresh.xml:1]: refresh icon for button

File: gcp/app-engine/test_main.py [hunk 11]
```diff
@@ -55,108 +55,108 @@ def test_json_serial_int():
...
-    assert actual_query == expected_query, f"Expected query: {expected_query}, but got: {actual_query}"
-
-    # Check the response data
-    expected_response = [
-        {'id': 1, 'name': 'Tournament 1'},
-        {'id': 2, 'name': 'Tournament 2'}
-    ]
-    assert response.status_code == 200
-    assert json.loads(response.data) == expected_response
+# def test_get_tournaments_no_args(client, mocker):
```
Comment (cfredericks) [inline gcp/app-engine/test_main.py:57]: comment out tests for now until we migrate over to the new endpoints (now that i've removed the sportsdataio endpoints). will add back later

File: .github/workflows/android.yml [hunk 12]
```diff
@@ -28,6 +28,10 @@ jobs:
         cache: gradle
     - name: Grant execute permission for gradlew
       run: chmod +x gradlew
+    - name: Load Google Service file
```
Comment (cfredericks) [inline .github/workflows/android.yml:31]: i added a github secret for running android CI (see comment in .gitignore for doing this locally)

File: .github/workflows/refresh_pga_data_function.yml [hunk 13]
```diff
@@ -18,19 +18,19 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install -r gcp/cloud-functions/refresh-leaderboards/requirements.txt
+        pip install -r gcp/cloud-functions/refresh-pga-data/requirements.txt
```
Comment (cfredericks) [inline .github/workflows/refresh_pga_data_function.yml:21]: unrelated fixes

File: app/src/main/res/layout/fragment_leaderboard.xml (hunk 2 above)
Comment (cfredericks) [reply app/src/main/res/layout/fragment_leaderboard.xml:20]: image

File: app/src/main/java/com/golfbetterapp/golfbetter/clients/PgaAppEngineClient.java [hunk 14]
```diff
@@ -35,145 +33,162 @@
  */
 public class PgaAppEngineClient {
   public static final String TOURNAMENT_FORMAT = "{tournament}";
-  public static final String TOURNAMENTS_ENDPOINT = "https://stoked-depth-428423-j7.uc.r.appspot.com/api/v1/pga-tournaments";
-  public static final String LEADERBOARD_PLAYERS_ENDPOINT = "https://stoked-depth-428423-j7.uc.r.appspot.com/api/v1/pga-leaderboard-players?tournamentId=" + TOURNAMENT_FORMAT;
-  public static final String PLAYER_SCORECARD_ENDPOINT = "https://stoked-depth-428423-j7.uc.r.appspot.com/api/v1/pga-player-scorecards?tournamentId=" + TOURNAMENT_FORMAT;
+
+  public static final String APP_SERVER_URL = "https://stoked-depth-428423-j7.uc.r.appspot.com";
+  // public static final String APP_SERVER_URL = "http://10.0.2.2:8080"; // For local dev
```
Comment (cfredericks) [inline app/src/main/java/com/golfbetterapp/golfbetter/clients/PgaAppEngineClient.java:38]: we should make this easier and configurable at some point, but for now if you want your emulator to communicate with a local version of app-engine running, you need to switch these values

File: gcp/app-engine/main.py [hunk 15]
```diff
@@ -7,15 +8,39 @@
...
 
+firebase_admin.initialize_app()
+
 DEFAULT_PROJECT_ID = 'stoked-depth-428423-j7'
 DEFAULT_VERSION_ID = 'latest'
 
 PUBLIC_API_PREFIX = '/api/v1'
 PRIVATE_API_PREFIX = '/protected/api/v1'
 
+# Decorator to parse auth token and extract user email
```
Comment (cfredericks) [inline gcp/app-engine/main.py:23]: decorator to (1) parse auth header, (2) throw 401 if missing or not valid, and (3) extract the user_email field and pass it into the flask callback

File: gcp/app-engine/main.py (hunk 6 above)
Comment (cfredericks) [reply gcp/app-engine/main.py]: NOTE: this has been added to this MR in a later commit

File: gcp/cloud-functions/refresh-pga-data/main.py [hunk 16]
```diff
@@ -779,10 +808,16 @@ def query_scorecard_v3(tournament_id, player_id):
                         help="Whether to write results to DB or only print results",
                         action=argparse.BooleanOptionalAction,
                         default=False)
+    parser.add_argument("-a",
```
Comment (cfredericks) [inline gcp/cloud-functions/refresh-pga-data/main.py:811]: add arg to backfill all tournaments (usually just reprocessed a whitelist passed in, or the current/previous tournament if no arg passed in. I ran this earlier today to populate all tournament data, which took well over an hourl

**Review by kevlaube — APPROVED** [review]
